{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.io\n",
    "import math\n",
    "import operator\n",
    "from time import time\n",
    "from Bio import SeqIO\n",
    "#more import xlsxwriter\n",
    "PRINT = True\n",
    "\n",
    "def mers(length):\n",
    "    \"\"\"Generates multimers for sorting through list of 10mers based on user\n",
    "    specification. Multimers generated act as the keys for generating a\n",
    "    hashtable to eliminate undesired sequence patterns from those 10mers not\n",
    "    found in the genome.\n",
    "\n",
    "    Usage: mers(N) = 4^(N) unique Nmers\n",
    "    \"\"\"\n",
    "    seq_list = ['']\n",
    "    counter=0\n",
    "    while counter < length:\n",
    "        for seq in seq_list:\n",
    "            if len(seq) == counter:\n",
    "                for x in ['A', 'T', 'C', 'G']:\n",
    "                    seq_list.append(seq+x)\n",
    "        counter += 1\n",
    "    last_N_Mers = 4**length\n",
    "    return seq_list[len(seq_list)-last_N_Mers :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identifyNucleotidePositionsOfMers(full_sequence, length=10):\n",
    "    \"\"\"Saves list of nucleotide positions in genome that all match a unique N-mer\n",
    "    sequence. Counting begins at __ending_ of MER.\n",
    "\n",
    "    Usage:   genomePositionsAtMers[mer_sequence] is a list of nucleotide positions\n",
    "    within the inputted fullSequence that match mer_sequence\n",
    "             If mer_sequence ends in a PAM site, then this can be used to match\n",
    "             the first N-3 nt of a guide strand plus a PAM site sequence.\n",
    "    \"\"\"\n",
    "\n",
    "    #Create a list of all possible N-mers\n",
    "    all_possible_mers = mers(length)\n",
    "\n",
    "    #Search through the genome and add nucleotide positions for match to an N-mer\n",
    "    positions_at_mers = {}\n",
    "    for mer in all_possible_mers:\n",
    "        positions_at_mers[mer] = []\n",
    "\n",
    "    print(\"Number of Mers: \", len(list(positions_at_mers.keys())))\n",
    "    counter = 0\n",
    "    while counter < (len(full_sequence)-length):\n",
    "        word = full_sequence[counter : counter+length]\n",
    "        positions_at_mers[word].append(counter+length)\n",
    "        counter += 1\n",
    "    return positions_at_mers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identifyTargetSequencesMatchingPAM(PAM_seq, positions_at_mers, full_sequence,\n",
    "                                        target_sequence_length=20):\n",
    "    \"\"\" Generates a list of target nucleotide sequences and corresponding nt\n",
    "    positions for an inputted sequence that matched the PAM_seq.\n",
    "        Uses the positionsAtMers dictionary to accelerate the identification.\n",
    "        Good for large genomes.\n",
    "\n",
    "    Usage:  listOfTargets = identifyTargetSequencesMatchingPAM('CGG',\n",
    "                                positionsAtMers, genome_sequence)\n",
    "    \"\"\"\n",
    "    target_sequence_list = []\n",
    "    all_mers = list(positions_at_mers.keys())\n",
    "    mer_length = len(all_mers[0])\n",
    "    list_of_mers_with_PAM = [mer + PAM_seq for mer in mers(mer_length - len(PAM_seq))]\n",
    "    for mer_with_PAM in list_of_mers_with_PAM:\n",
    "        nt_list = positions_at_mers[mer_with_PAM]\n",
    "        for nt in nt_list:\n",
    "            begin = nt-target_sequence_length - len(PAM_seq)\n",
    "            end = nt - len(PAM_seq)\n",
    "            if begin > 0 and end < len(full_sequence): #Does not account for circular DNAs\n",
    "                target_sequence = full_sequence[begin : end]\n",
    "                target_sequence_list.append((target_sequence, nt))\n",
    "    return target_sequence_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "class clCas9Calculator(object):\n",
    "\n",
    "    def __init__(self,filename, quickmode=True, ModelName='InvitroModel.mat'):\n",
    "\n",
    "        self.quickmode=quickmode\n",
    "        self.ModelName=ModelName\n",
    "        data=scipy.io.loadmat(self.ModelName)\n",
    "        self.weights=data['w1']\n",
    "        self.decNN=data['decNN']\n",
    "        self.RT = 0.61597\n",
    "\n",
    "        # the PAMs with the highest dG, ignoring other PAM sequences by setting their dG to 0\n",
    "        self.PAM_energy={'GGA':-9.8,'GGT':-10,'GGC':-10,'GGG':-9.9,'CGG':-8.1,'TGG':-7.8,'AGG':-8.1,'AGC':-8.1,'AGT':-8.1,'AGA':-7.9,'GCT':-7.1,'GCG':-6.9,'ATT':-7.2,'ATC':-6.4,'TTT':-7.6,'TTG':-6.8,'GTA':-7.4,'GTT':-7.9,'GTG':-7.7,'AAT':-7,'AAG':-7,'TAT':-7.2,'TAG':-7.2,'GAA':-7.2,'GAT':-7.3,'GAC':-7.2,'GAG':-7.3}\n",
    "\n",
    "        self.initGenomeFinder(filename)\n",
    "\n",
    "    def returnAllPAMs(self):\n",
    "\n",
    "        for (PAMpart,energies) in sorted(list(self.PAM_energy.items()), key = lambda x: x[1]):  #PAMpart will be 'GGT'\n",
    "            for nt in ('A','G','C','T'):        #nt + PAMpart will be all possible 'NGGT'\n",
    "                yield nt + PAMpart\n",
    "\n",
    "    def initGenomeFinder(self, filename):\n",
    "\n",
    "        targetDictionary = {}\n",
    "\n",
    "        handle = open(filename,'r')\n",
    "        records = SeqIO.parse(handle,\"fasta\")\n",
    "        record = next(records)\n",
    "        handle.close()\n",
    "\n",
    "        fullSequence = str(record.seq)\n",
    "        print(\"full seq length\", len(record.seq))\n",
    "        positionsAtMers = identifyNucleotidePositionsOfMers(fullSequence, length = 10)\n",
    "        print(\"computed positions\")\n",
    "        targetDictionary[filename] = {}\n",
    "        targetSequenceList = []\n",
    "        for fullPAM in self.returnAllPAMs():\n",
    "            print(\"Full PAM\", fullPAM)\n",
    "            targetSequenceList = identifyTargetSequencesMatchingPAM(fullPAM, positionsAtMers, fullSequence)\n",
    "            targetDictionary[filename][fullPAM] = targetSequenceList\n",
    "        self.targetDictionary = targetDictionary\n",
    "\n",
    "    def printModelInfo(self):\n",
    "        m=0\n",
    "        s=0\n",
    "        negative_val=0\n",
    "        for i,l in enumerate(self.decNN):\n",
    "            for j,e in enumerate(l):\n",
    "                if float(e)<0:\n",
    "                    negative_val+=1\n",
    "                if i!=j:\n",
    "                    s+=float(e)\n",
    "                    m+=1\n",
    "        meanNN=float(s)/float(m)\n",
    "\n",
    "        sw=0\n",
    "        for w in self.weights:\n",
    "            sw+=w\n",
    "\n",
    "        meanw=sw/len(self.weights)\n",
    "        print('average mismatchc energy: ', meanNN)\n",
    "        print('average weight:', meanw)\n",
    "        print('number of negative energies: ', negative_val)\n",
    "\n",
    "    def Calc_Exchange_Energy(self,crRNA,targetSeq):\n",
    "        nt_pos={'A':0,'T':1,'C':2,'G':3,'a':0,'t':1,'c':2,'g':3}\n",
    "        dG=0\n",
    "        RNA=''\n",
    "        DNA=''\n",
    "        for i in range(0,len(crRNA)):\n",
    "            if i>0:\n",
    "                RNA=crRNA[(i-1):(i+1)]\n",
    "                DNA=targetSeq[(i-1):(i+1)]\n",
    "                RNA_index=nt_pos[RNA[0]]+4*nt_pos[RNA[1]]\n",
    "                DNA_index=nt_pos[DNA[0]]+4*nt_pos[DNA[1]]\n",
    "\n",
    "                dG1=float(self.decNN[RNA_index][DNA_index])\n",
    "                if abs(dG1-0.000015)<1e-6:\n",
    "                    dG1=10000\n",
    "                    dG1=2.3 # during model identification, I set the value of every unknown dG to 0.000015 (if I did not find a value for it)\n",
    "\n",
    "\n",
    "                pos=20-i\n",
    "                w1=float(self.weights[pos])\n",
    "                #print 'b1',RNA[0],RNA[1],DNA[0],DNA[1],RNA_index, DNA_index, pos,dG1, w1\n",
    "            else:\n",
    "                w1=0\n",
    "                dG1=0\n",
    "            if i<(len(crRNA)-1):\n",
    "                RNA2=crRNA[i:(i+2)]\n",
    "                DNA2=targetSeq[i:(i+2)]\n",
    "                RNA_index=nt_pos[RNA2[0]]+4*nt_pos[RNA2[1]]\n",
    "                DNA_index=nt_pos[DNA2[0]]+4*nt_pos[DNA2[1]]\n",
    "                dG2=float(self.decNN[RNA_index][DNA_index])\n",
    "                if abs(dG2-0.000015)<1e-6:\n",
    "                    dG2=10000\n",
    "                    dG2=2.3 # during model identification, I set the value of every unknown dG to 0.000015 (if I did not find a value for it)\n",
    "\n",
    "                pos=20-i-1\n",
    "                w2=float(self.weights[pos])\n",
    "                #print 'b2',RNA2[0],RNA2[1],DNA2[0],DNA2[1],RNA_index, DNA_index, pos,dG2, w2\n",
    "            else:\n",
    "                w2=0\n",
    "                dG2=0\n",
    "            dG+=w1*dG1+w2*dG2\n",
    "        return float(dG)\n",
    "\n",
    "    def QuickCalc_Exchange_Energy(self,crRNA,TargetSeq):\n",
    "        nt_pos={'A':0,'T':1,'C':2,'G':3,'a':0,'t':1,'c':2,'g':3}\n",
    "        dG=0\n",
    "        RNA=''\n",
    "        DNA=''\n",
    "        self.nt_mismatch_in_first8=0\n",
    "        for i in range(0,len(crRNA)):\n",
    "            pos=20-i\n",
    "            w1=self.weights[pos]\n",
    "            if nt_pos[crRNA[i]]==nt_pos[TargetSeq[i]]:\n",
    "                dG1=0\n",
    "            else:\n",
    "                # using a bioinformatics search approach to find sequences with up to x mismatches\n",
    "                dG1=2.3 # kcal/mol\n",
    "                if pos<=8:\n",
    "                    self.nt_mismatch_in_first8=self.nt_mismatch_in_first8+1\n",
    "            dG+=w1*dG1\n",
    "        return float(dG)\n",
    "\n",
    "    def calc_dG_PAM(self, PAM_full_seq):\n",
    "\n",
    "        #PAM sequence is 5' - N xxx N - 3' where the energy of xxx is listed below. A normal PAM of 'NGG' with 'TC' afterwards would be listed as 'GGT'\n",
    "        key = PAM_full_seq[1:4]\n",
    "        if key in self.PAM_energy:\n",
    "            return self.PAM_energy[key]\n",
    "        else:\n",
    "            return 0.0\n",
    "\n",
    "        # acceptedPAMList=PAM_dic_energy.keys()\n",
    "        # self.dG_PAM_List=[]\n",
    "        # self.WarningPAM_List=[]\n",
    "        # PAMsize=len(self.PAM)\n",
    "        # for target in self.sequence_list:\n",
    "            # tPAM=target[-(PAMsize):-1]+target[-1]\n",
    "            # if tPAM in acceptedPAMList:\n",
    "                # dGPAM=PAM_dic_energy[tPAM]\n",
    "                # warning=''\n",
    "            # else:\n",
    "                # dGPAM=0\n",
    "                # warning='N.B'\n",
    "            # self.dG_PAM_List.append(dGPAM)\n",
    "            # self.WarningPAM_List.append(warning)\n",
    "\n",
    "    def calc_dG_exchange(self, guideSequence, targetSequence):\n",
    "        self.nt_mismatch_in_first8_list=[]\n",
    "        if self.quickmode:\n",
    "            solverfunc=self.QuickCalc_Exchange_Energy\n",
    "        else:\n",
    "            solverfunc=self.Calc_Exchange_Energy\n",
    "\n",
    "        dG_exchange = solverfunc(guideSequence,targetSequence)\n",
    "\n",
    "        return dG_exchange\n",
    "\n",
    "    def calc_dG_supercoiling(self, sigmaInitial, targetSequence):\n",
    "\n",
    "\n",
    "        sigmaFinal = -0.08\n",
    "        dG_supercoiling = 10.0 * len(targetSequence) * self.RT * (sigmaFinal**2 - sigmaInitial**2)\n",
    "        return dG_supercoiling\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "full seq length 4653140\n",
      "Number of Mers:  1048576\n",
      "computed positions\n",
      "Full PAM AGGT\n",
      "Full PAM GGGT\n",
      "Full PAM CGGT\n",
      "Full PAM TGGT\n",
      "Full PAM AGGC\n",
      "Full PAM GGGC\n",
      "Full PAM CGGC\n",
      "Full PAM TGGC\n",
      "Full PAM AGGG\n",
      "Full PAM GGGG\n",
      "Full PAM CGGG\n",
      "Full PAM TGGG\n",
      "Full PAM AGGA\n",
      "Full PAM GGGA\n",
      "Full PAM CGGA\n",
      "Full PAM TGGA\n",
      "Full PAM ACGG\n",
      "Full PAM GCGG\n",
      "Full PAM CCGG\n",
      "Full PAM TCGG\n",
      "Full PAM AAGG\n",
      "Full PAM GAGG\n",
      "Full PAM CAGG\n",
      "Full PAM TAGG\n",
      "Full PAM AAGC\n",
      "Full PAM GAGC\n",
      "Full PAM CAGC\n",
      "Full PAM TAGC\n",
      "Full PAM AAGT\n",
      "Full PAM GAGT\n",
      "Full PAM CAGT\n",
      "Full PAM TAGT\n",
      "Full PAM AAGA\n",
      "Full PAM GAGA\n",
      "Full PAM CAGA\n",
      "Full PAM TAGA\n",
      "Full PAM AGTT\n",
      "Full PAM GGTT\n",
      "Full PAM CGTT\n",
      "Full PAM TGTT\n",
      "Full PAM ATGG\n",
      "Full PAM GTGG\n",
      "Full PAM CTGG\n",
      "Full PAM TTGG\n",
      "Full PAM AGTG\n",
      "Full PAM GGTG\n",
      "Full PAM CGTG\n",
      "Full PAM TGTG\n",
      "Full PAM ATTT\n",
      "Full PAM GTTT\n",
      "Full PAM CTTT\n",
      "Full PAM TTTT\n",
      "Full PAM AGTA\n",
      "Full PAM GGTA\n",
      "Full PAM CGTA\n",
      "Full PAM TGTA\n",
      "Full PAM AGAT\n",
      "Full PAM GGAT\n",
      "Full PAM CGAT\n",
      "Full PAM TGAT\n",
      "Full PAM AGAG\n",
      "Full PAM GGAG\n",
      "Full PAM CGAG\n",
      "Full PAM TGAG\n",
      "Full PAM AATT\n",
      "Full PAM GATT\n",
      "Full PAM CATT\n",
      "Full PAM TATT\n",
      "Full PAM ATAT\n",
      "Full PAM GTAT\n",
      "Full PAM CTAT\n",
      "Full PAM TTAT\n",
      "Full PAM ATAG\n",
      "Full PAM GTAG\n",
      "Full PAM CTAG\n",
      "Full PAM TTAG\n",
      "Full PAM AGAA\n",
      "Full PAM GGAA\n",
      "Full PAM CGAA\n",
      "Full PAM TGAA\n",
      "Full PAM AGAC\n",
      "Full PAM GGAC\n",
      "Full PAM CGAC\n",
      "Full PAM TGAC\n",
      "Full PAM AGCT\n",
      "Full PAM GGCT\n",
      "Full PAM CGCT\n",
      "Full PAM TGCT\n",
      "Full PAM AAAT\n",
      "Full PAM GAAT\n",
      "Full PAM CAAT\n",
      "Full PAM TAAT\n",
      "Full PAM AAAG\n",
      "Full PAM GAAG\n",
      "Full PAM CAAG\n",
      "Full PAM TAAG\n",
      "Full PAM AGCG\n",
      "Full PAM GGCG\n",
      "Full PAM CGCG\n",
      "Full PAM TGCG\n",
      "Full PAM ATTG\n",
      "Full PAM GTTG\n",
      "Full PAM CTTG\n",
      "Full PAM TTTG\n",
      "Full PAM AATC\n",
      "Full PAM GATC\n",
      "Full PAM CATC\n",
      "Full PAM TATC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<__main__.clCas9Calculator at 0xb2055abe0>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clCas9Calculator(\"Fasta_Files/E_coli_MG1655_genome.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Azimuth_Model import model_comparison\n",
    "Guide_RNAs = [\"CCAGAAGTTTGAGCCACAAACCCATGGTCA\", \"AAGTTAACTTCAAAATTAGACACAACATTG\"]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/sklearn/base.py:253: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.20.2 when using version 0.20.3. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n",
      "/anaconda3/lib/python3.7/site-packages/sklearn/base.py:253: UserWarning: Trying to unpickle estimator GradientBoostingRegressor from version 0.20.2 when using version 0.20.3. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n"
     ]
    }
   ],
   "source": [
    "predictions = model_comparison.predict(np.array(Guide_RNAs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.64524886])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
